## Identification of mPing/Ping/Pong excision

+ Working directory on biocluster: /rhome/cjinfeng/BigData/00.RD/RILs/Transpostion/bin/Manuscript_Preparation/Question11_fix_excision_module/.

+ Create pseudogenomes with mPing/Ping/Pong inserted/excised from the reference genome and align paired-end reads to these genomes.

```shell
echo "Pseudogenome with 507 mPing and 7 Ping inserted"
python Pseudo_TEinsertion_Genome.py --repeat mPing_Ping_Pong.fa --gff RILs.ALL_mPing_Ping.gff --genome MSU_r7.fa --project MSU_r7.Pseudo_mPing_Ping_514

arguments:
    --repeat: fasta format sequence of mPing/Ping/Pong sequence
    --gff: gff file of 514 nonreference mPing/Ping insertions
    --genome: the reference genome (MSUv7) to be inserted
    --project: prefix of output file

python runRIL_bwa.py --input single_ping_rils_fastq --ref MSU_r7.Pseudo_mPing_Ping_514.fa

arguments:
    --input: directory of fastq files
    --ref: pseudogenomes used as reference to align the reads 

echo "Pseudogenome with 51 mPing and 1 Ping and 5 Pong excised"
python Pseudo_TEinsertion_Genome_Ref.py --gff Parent.ALL.mPing_Ping_Pong.Ref_Shared.gff --genome MSU_r7.fa --project Parent.Pseudo_mPing_Ping_Pong_57.Ref_Shared
python runRIL_bwa.py --input single_ping_rils_fastq --ref Parent.Pseudo_mPing_Ping_Pong_57.Ref_Shared.fa 
```

+ Analyze read coverage (covered: reads cover junction of TE insertions; clipped: reads have softclip at junction of TE insertions) at mPing/Ping/Pong insertions using bam files with reads aligned to the reference genome and pseudogenome.

```shell
echo "Nonreference mPing/Ping, including 415 mPing and 7 Ping in HEG4 plus 92 other mPings that are present in many RILs (allele frequency > 0.1). Ping16A_Stow (PingA) is processed by a different script."
python mPing_Boundary_Coverage_MP.py --bam_ref RILs_ALL_bam_correct_merged --bam_pseudo RILs_ALL_unmapped_mping_bam_RILs_514_mPing_Ping --gff_ref RILs.ALL_mPing_Ping.gff --gff_pseudo MSU_r7.Pseudo_mPing_Ping_514.gff --project output_MSU7_nonreference_514_MP3

arguments:
    --bam: directory that contains bam files with reads aligned to the reference genome
    --bam_pseudo: directory that has bam files with reads aligned to the pseudogenome with mPing/Ping/Pong inserted in the reference genome
    --gff: gff file of nonreference mPing/Ping insertions in the reference genome
    --gff_pseudo: gff file of mPing/Ping insertions (become reference insertions) in the pseudogenome
    --project: prefix for the output files

echo "Reference and shared mPing/Ping, including 44 reference and 7 shared mPing plus 1 reference Ping and 5 reference Pong"
python mPing_Boundary_Coverage_Ref_MP.py --bam_ref RILs_ALL_unmapped_mping_bam_Ref_57_mPing_Ping_Pong --bam_pseudo RILs_ALL_bam_correct_merged/ --gff_ref Parent.Pseudo_mPing_Ping_Pong_57.Ref_Shared.gff --gff_pseudo Parent.ALL.mPing_Ping_Pong.Ref_Shared.gff --project output_MSU7_reference_57_MP3

arguments:
    --bam: directory that contains bam files with reads aligned to the pseudogenome with mPing/Ping/Pong removed from the reference genome
    --bam_pseudo: directory that has bam files with reads aligned to the reference genome
    --gff: gff file of reference mPing/Ping insertions in the pseudogenome
    --gff_pseudo: gff file of mPing/Ping insertions in the reference genome
    --project: prefix for the output files
```

+ Determine mPing/Ping/Pong excisions based on read coverage at mPing/Ping/Pong insertions.

```shell
echo "Nonreference mPing/Ping/Pong"
python Ping_number_RILs.High_exicison.py --csv output_MSU7_nonreference_514_MP3_mPing --ping_code RIL272_RelocaTEi.Jinfeng_Lulu.ping_code.table.txt --output output_MSU7_nonreference_514_MP3_mPing_GT_Ping_code

arguments:
    --csv: directory that contains read coverage of mPing/Ping insertions generated by "mPing_Boundary_Coverage_MP.py"
    --ping_code: table with Ping copy number and Ping combination code for each RILs
    --output: output directory

echo "Reference and shared mPing/Ping"
python Ping_number_RILs.High_exicison_Ref.py --csv output_MSU7_reference_57_MP3_mPing --ping_code RIL272_RelocaTEi.Jinfeng_Lulu.ping_code.table.txt --gff Parent.ALL.mPing_Ping_Pong.Ref_Shared.gff --output output_MSU7_reference_57_MP3_mPing_GT_Ping_code

arguments:
    --csv: directory that contains read coverage of mPing/Ping insertions generated by "mPing_Boundary_Coverage_Ref_MP.py"
    --ping_code: table with Ping copy number and Ping combination code for each RILs
    --output: output directory
```

+ Generate list of RILs that have excisions for each mPing/Ping/Pong loci

```shell
echo "Nonreference mPing/Ping/Pong"
cd output_MSU7_nonreference_514_MP3_mPing_GT_Ping_code
#These four loci have name issues. We create links with their correct names.
ln -s Chr3_28019798_28019800.matrix.csv Chr3_28019800_28019802.matrix.csv
ln -s Chr9_10863116_10863118.matrix.csv Chr9_10863118_10863120.matrix.csv
ln -s Chr9_13736139_13736141.matrix.csv Chr9_13736141_13736143.matrix.csv
ln -s Chr9_16690610_16690612.matrix.csv Chr9_16690612_16690614.matrix.csv
cd ..
python Sum_mPing_excision.py --dir output_MSU7_nonreference_514_MP3_mPing_GT_Ping_code --gff Parent.ALL.mPing_Ping.422.gff

arguments:
    --dir: directory that contains excision status of each nonreference mPing/Ping in each RILs
    --gff: gff file of nonreference mPing/Ping loci to be processed

echo "Reference and shared mPing/Ping"
python Sum_mPing_excision_Ref.py --dir output_MSU7_reference_57_MP3_mPing_GT_Ping_code/ --gff Parent.ALL.mPing_Ping_Pong.Ref_Shared.gff

arguments:
    --dir: directory that contains excision status of each reference mPing/Ping/Pong in each RILs
    --gff: gff file of reference and shared mPing/Ping/Pong loci to be processed

echo "Select random loci with excisions to validate by PCR"
python Random_sample_validation.py --input 30 > Random_sample_validation.loci.list.txt
```

+ Characterize excision footprints

```shell
echo "Nonreference mPing/Ping/Pong"
python footprint_events.py --input output_MSU7_nonreference_514_MP3_mPing_GT_Ping_code.mping_excision.list 

arguments:
   --input: list of RILs that have excisions for each nonreference mPing/Ping loci

echo "Reference and shared mPing/Ping"
python footprint_events_Ref.py --input output_MSU7_reference_57_MP3_mPing_GT_Ping_code.mping_excision.list

arguments:
   --input: list of RILs that have excisions for each reference and shared mPing/Ping/Pong loci 
```





